airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/ctid/InitialSyncCtidIterator.java

```diff
@@ -0,0 +1,230 @@
+import io.airbyte.integrations.source.postgres.PostgresQueryUtils;
+import io.airbyte.integrations.source.postgres.ctid.CtidPostgresSourceOperations.RowDataWithCtid;
+import io.airbyte.integrations.source.postgres.internal.models.CtidStatus;
+import io.airbyte.integrations.source.relationaldb.RelationalDbQueryUtils;
+import io.airbyte.protocol.models.AirbyteStreamNameNamespacePair;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.stream.Stream;
+import javax.annotation.CheckForNull;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class InitialSyncCtidIterator extends AbstractIterator<RowDataWithCtid> implements AutoCloseableIterator<RowDataWithCtid> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(InitialSyncCtidIterator.class);
+  private static final int MAX_ALLOWED_RESYNCS = 5;
+  private static final int QUERY_TARGET_SIZE_GB = 1;
+  private static final double MEGABYTE = Math.pow(1024, 2);
+  private static final double GIGABYTE = MEGABYTE * 1024;
+
+  private final AirbyteStreamNameNamespacePair airbyteStream;
+  private final long blockSize;
+  private final List<String> columnNames;
+  private final CtidStateManager ctidStateManager;
+  private final JdbcDatabase database;
+  private final FileNodeHandler fileNodeHandler;
+  private final String quoteString;
+  private final String schemaName;
+  private final CtidPostgresSourceOperations sourceOperations;
+  private final Queue<Pair<Ctid, Ctid>> subQueriesPlan;
+  private final String tableName;
+  private final long tableSize;
+  private AutoCloseableIterator<RowDataWithCtid> currentIterator;
+
+  private Long lastKnownFileNode;
+  private int numberOfTimesReSynced = 0;
+
+  public InitialSyncCtidIterator(final CtidStateManager ctidStateManager,
+      final JdbcDatabase database,
+      final CtidPostgresSourceOperations sourceOperations,
+      final String quoteString,
+      final List<String> columnNames,
+      final String schemaName,
+      final String tableName,
+      final long tableSize,
+      final long blockSize,
+      final FileNodeHandler fileNodeHandler) {
+    this.airbyteStream = AirbyteStreamUtils.convertFromNameAndNamespace(tableName, schemaName);
+    this.blockSize = blockSize;
+    this.columnNames = columnNames;
+    this.ctidStateManager = ctidStateManager;
+    this.database = database;
+    this.fileNodeHandler = fileNodeHandler;
+    this.quoteString = quoteString;
+    this.schemaName = schemaName;
+    this.sourceOperations = sourceOperations;
+    this.subQueriesPlan = new LinkedList<>();
+    this.tableName = tableName;
+    this.tableSize = tableSize;
+  }
+
+  @CheckForNull
+  @Override
+  protected RowDataWithCtid computeNext() {
```

Comment:
I know this is not trivial, but if we could test the logic in this `computeNext()` class since it's very stateful.
Compare to CompositeIteratorTest